osutil/kcmdline_test.go

```diff
@@ -208,3 +208,70 @@ func (s *kcmdlineTestSuite) TestKernelCommandLine(c *C) {
 	c.Assert(err, IsNil)
 	c.Check(cmd, Equals, "foo bar baz panic=-1")
 }
+
+// Outcome of test cases here match the next_arg function from
+// lib/cmdline.c in the linux kernel (minus sometimes the linux code
+// returning NULL or "" while we always return "" in both cases, but
+// that should be irrelevant).
+func (s *kcmdlineTestSuite) TestKernelParseCommandLine(c *C) {
+	for idx, tc := range []struct {
+		cmd string
+		exp []osutil.KernelArgument
+	}{
+		{cmd: ``, exp: []osutil.KernelArgument{}},
+		{cmd: `foo bar baz`, exp: []osutil.KernelArgument{
+			{"foo", ""}, {"bar", ""}, {"baz", ""}}},
+		{cmd: `"foo"=" many   spaces  " bar`, exp: []osutil.KernelArgument{
+			{`foo"`, " many   spaces  "}, {"bar", ""}}},
+		{cmd: `"foo=bar" foo="bar"`, exp: []osutil.KernelArgument{
+			{"foo", "bar"}, {"foo", "bar"}}},
+		{cmd: `foo_bar-tee=bar_aa-bb`, exp: []osutil.KernelArgument{
+			{"foo_bar_tee", "bar_aa-bb"}}},
+		{cmd: `foo="1$2"`, exp: []osutil.KernelArgument{{"foo", "1$2"}}},
+		{cmd: `foo=1$2`, exp: []osutil.KernelArgument{{"foo", "1$2"}}},
+		{cmd: `foo= bar`, exp: []osutil.KernelArgument{{"foo", ""}, {"bar", ""}}},
+		{cmd: `foo=""`, exp: []osutil.KernelArgument{{"foo", ""}}},
+		{cmd: `   cpu=1,2,3   mem=0x2000;0x4000:$2  `,
+			exp: []osutil.KernelArgument{{"cpu", "1,2,3"}, {"mem", "0x2000;0x4000:$2"}}},
+		{cmd: "isolcpus=1,2,10-20,100-2000:2/25",
+			exp: []osutil.KernelArgument{{"isolcpus", "1,2,10-20,100-2000:2/25"}}},
+		// something more realistic
+		{
+			cmd: `BOOT_IMAGE=/vmlinuz-linux root=/dev/mapper/linux-root rw quiet loglevel=3 rd.udev.log_priority=3 vt.global_cursor_default=0 rd.luks.uuid=1a273f76-3118-434b-8597-a3b12a59e017 rd.luks.uuid=775e4582-33c1-423b-ac19-f734e0d5e21c rd.luks.options=discard,timeout=0 root=/dev/mapper/linux-root apparmor=1 security=apparmor`,
+			exp: []osutil.KernelArgument{
+				{"BOOT_IMAGE", "/vmlinuz-linux"},
+				{"root", "/dev/mapper/linux-root"},
+				{"rw", ""},
+				{"quiet", ""},
+				{"loglevel", "3"},
+				{"rd.udev.log_priority", "3"},
+				{"vt.global_cursor_default", "0"},
+				{"rd.luks.uuid", "1a273f76-3118-434b-8597-a3b12a59e017"},
+				{"rd.luks.uuid", "775e4582-33c1-423b-ac19-f734e0d5e21c"},
+				{"rd.luks.options", "discard,timeout=0"},
+				{"root", "/dev/mapper/linux-root"},
+				{"apparmor", "1"},
+				{"security", "apparmor"},
+			},
+		},
+		// this is actually ok, eg. rd.luks.options=discard,timeout=0
+		{cmd: `a=b=`, exp: []osutil.KernelArgument{{"a", "b="}}},
+		// bad quoting, or otherwise malformed command line
+		{cmd: `foo="1$2`, exp: []osutil.KernelArgument{{"foo", "1$2"}}},
+		{cmd: `"foo"`, exp: []osutil.KernelArgument{{"foo", ""}}},
+		{cmd: `foo"foo"`, exp: []osutil.KernelArgument{{`foo"foo"`, ""}}},
+		{cmd: `foo=foo"`, exp: []osutil.KernelArgument{{"foo", `foo"`}}},
```

Comment:
what happens with foo=bar=bat or with extra " in the cmdline?

Reply:
`foo=bar=bat` -> splits to `foo`, `bar=bat` (added test case). There are already some test cases with quotes in different places, which specific test are you thinking about? Note that the behavior of the kernel is a bit peculiar, quoting the full argument (`"par=val"`) or quoting just the value (`par="val"`) is fine, but escaping quotes is not allowed, and it won't clean-up quotes in the middle for instance. So it is not like any random string can be used as param/value.

Reply:
thinking about something like  "f"o"o"="b"a"r", what does that procude? does it match the kernel behavior?